pulsar-io/nifi/src/main/java/org/apache/pulsar/io/nifi/NiFiSink.java

```diff
@@ -0,0 +1,141 @@
+    private List<Record<NiFiDataPacket>>  currentList;
+    private ScheduledExecutorService flushExecutor;
+
+    @Override
+    public void open(Map<String, Object> config, SinkContext sinkContext) throws Exception {
+        niFiConfig = NiFiConfig.load(config);
+        Preconditions.checkNotNull(niFiConfig.getUrl(), "url property not set.");
+        Preconditions.checkNotNull(niFiConfig.getPortName(), "portName property not set.");
+        Preconditions.checkArgument(niFiConfig.getRequestBatchCount() > 0,
+                "requestBatchCount must be a positive integer.");
+
+        clientConfig = new SiteToSiteClient.Builder()
+                .url(niFiConfig.getUrl())
+                .portName(niFiConfig.getPortName())
+                .requestBatchCount(niFiConfig.getRequestBatchCount())
+                .buildConfig();
+        client = new SiteToSiteClient.Builder().fromConfig(clientConfig).build();
+
+
+        requestBatchCount = niFiConfig.getRequestBatchCount();
+        waitTimeMs = niFiConfig.getWaitTimeMs();
+        currentList= Lists.newArrayList();
+
+        flushExecutor = Executors.newScheduledThreadPool(1);
+        flushExecutor.scheduleAtFixedRate(() -> flush(), waitTimeMs, waitTimeMs, TimeUnit.MILLISECONDS);
+    }
+
+    @Override
+    public void close() throws Exception {
+        if (null != client) {
+            client.close();
+        }
+
+        if (null != flushExecutor) {
+            flushExecutor.shutdown();
+        }
+
+    }
+
+    @Override
+    public void write(Record<NiFiDataPacket> record) {
+        int number;
+        synchronized (this) {
+            currentList.add(record);
+            number = currentList.size();
+        }
+
+        if (number == requestBatchCount) {
+            flushExecutor.schedule(() -> flush(), 0, TimeUnit.MILLISECONDS);
+        }
+    }
+
+    private void flush() {
+        try {
+            final Transaction transaction = client.createTransaction(TransferDirection.SEND);
+            List<Record<NiFiDataPacket>>  toFlushList;
+            synchronized (this) {
+                if (currentList.isEmpty()) {
+                    return;
+                }
+                toFlushList = currentList;
+                currentList = Lists.newArrayList();
+            }
+
+            for (Record<NiFiDataPacket> record : toFlushList) {
+                NiFiDataPacket niFiDataPacket = record.getValue();
+                transaction.send(niFiDataPacket.getContent(), niFiDataPacket.getAttributes());
+            }
+            try {
+                transaction.confirm();
+                transaction.complete();
+                toFlushList.forEach(record -> record.ack());
+            } catch (Exception e) {
+                log.warn("Record flush thread was interrupted", e);
+                toFlushList.forEach(record -> record.fail());
+            }
+        } catch (final IOException ioe) {
+            log.warn("Failed to receive data from NiFi", ioe);
```

Comment:
who throw this exception? is it `client.createTransaction`? if so, don't we need to handle failing the records in toFlushList? Otherwise, these records are not processed, right?